File: LC1_TwoSum/HashTable/Hash.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // Hash Table
        // Time Complexty: O(n)
        // Space Complexty: O(n)
        unordered_map<int, int> mapping;

        for(int i = 0; i < nums.size(); i++){
            mapping[nums[i]] = i;
        }

        vector<int> result;
        for(int i = 0; i < nums.size(); i++){
            if(mapping.count(target - nums[i]) == 1 && mapping[target - nums[i]] != i){
                result.push_back(i);
                result.push_back(mapping[target - nums[i]]);
                break;
            }
        }

        return result;
    }
};

File: LC128_LongestConsecutiveSequence/hash.py
from typing import List


class Solution:
    # Hash Set
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def longestConsecutive(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        
        hashSet = set(nums)
        bestSoFar = 1

        for num in nums:
            if num-1 not in hashSet:
                temp = num + 1
                count = 1
                while temp in hashSet:
                    count += 1
                    temp += 1
                bestSoFar = max(bestSoFar, count)
        return bestSoFar